From bed587268470de040ae7589295682ed38f80f87e Mon Sep 17 00:00:00 2001
From: Satish Patel <tosatishpatel@gmail.com>
Date: Thu, 29 Nov 2018 11:07:49 +0530
Subject: [PATCH] Settings: Add developer setting for root access

Also includes following change:

    Settings: Set root access options appropriately

    It is possible to be running a user build with a debuggable boot image.
    In this case, "su" will not be available.  So only show none/adb.

Change-Id: I9d7a3bf4e4748a8643f89e78bc07e23de89dde6f
Signed-off-by: Satish Patel <tosatishpatel@gmail.com>
---
 res/values/lineage_arrays.xml                 |  43 ++++++
 res/values/strings.xml                        |   9 ++
 res/xml/development_settings.xml              |   5 +
 .../DevelopmentSettingsDashboardFragment.java |  68 +++++----
 .../development/RootAccessDialogHost.java     |  33 +++++
 .../RootAccessPreferenceController.java       | 137 ++++++++++++++++++
 .../development/RootAccessWarningDialog.java  |  82 +++++++++++
 7 files changed, 351 insertions(+), 26 deletions(-)
 create mode 100644 res/values/lineage_arrays.xml
 create mode 100644 src/com/android/settings/development/RootAccessDialogHost.java
 create mode 100644 src/com/android/settings/development/RootAccessPreferenceController.java
 create mode 100644 src/com/android/settings/development/RootAccessWarningDialog.java

diff --git a/res/values/lineage_arrays.xml b/res/values/lineage_arrays.xml
new file mode 100644
index 0000000000..8e51beec66
--- /dev/null
+++ b/res/values/lineage_arrays.xml
@@ -0,0 +1,43 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+     Copyright (C) 2012-2015 The CyanogenMod Project
+     Copyright (C) 2018 The LinegeOS Project
+
+     Licensed under the Apache License, Version 2.0 (the "License");
+     you may not use this file except in compliance with the License.
+     You may obtain a copy of the License at
+
+          http://www.apache.org/licenses/LICENSE-2.0
+
+     Unless required by applicable law or agreed to in writing, software
+     distributed under the License is distributed on an "AS IS" BASIS,
+     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+     See the License for the specific language governing permissions and
+     limitations under the License.
+-->
+<resources>
+    <!-- Arrays for root access capability -->
+    <string-array name="root_access_entries" translatable="false">
+        <item>@string/root_access_none</item>
+        <item>@string/root_access_apps</item>
+        <item>@string/root_access_adb</item>
+        <item>@string/root_access_all</item>
+    </string-array>
+
+    <string-array name="root_access_values" translatable="false">
+        <item>0</item>
+        <item>1</item>
+        <item>2</item>
+        <item>3</item>
+    </string-array>
+
+    <string-array name="root_access_entries_adb" translatable="false">
+        <item>@string/root_access_none</item>
+        <item>@string/root_access_adb</item>
+    </string-array>
+
+    <string-array name="root_access_values_adb" translatable="false">
+        <item>0</item>
+        <item>2</item>
+    </string-array>
+</resources>
diff --git a/res/values/strings.xml b/res/values/strings.xml
index fb5f395571..1ae0b81e7d 100644
--- a/res/values/strings.xml
+++ b/res/values/strings.xml
@@ -10085,4 +10085,13 @@
 
     <!-- Title for Connected device shortcut [CHAR LIMIT=30] -->
     <string name="devices_title">Devices</string>
+
+	<!-- Setting checkbox title for root access -->
+    <string name="root_access">Root access</string>
+    <string name="root_access_warning_title">Allow root access?</string>
+    <string name="root_access_warning_message">Allowing apps to request root access is very dangerous and could compromise the security of your system!</string>
+    <string name="root_access_none">Disabled</string>
+    <string name="root_access_apps">Apps only</string>
+    <string name="root_access_adb">ADB only</string>
+    <string name="root_access_all">Apps and ADB</string>
 </resources>
diff --git a/res/xml/development_settings.xml b/res/xml/development_settings.xml
index 74f29b3ba7..1e59c12a05 100644
--- a/res/xml/development_settings.xml
+++ b/res/xml/development_settings.xml
@@ -119,6 +119,11 @@
             android:key="quick_settings_tiles"
             android:title="@string/quick_settings_developer_tiles"
             android:fragment="com.android.settings.development.qstile.DevelopmentTileConfigFragment" />
+
+           <ListPreference
+            android:key="root_access"
+            android:title="@string/root_access"
+            android:persistent="false" /> 
     </PreferenceCategory>
 
     <PreferenceCategory
diff --git a/src/com/android/settings/development/DevelopmentSettingsDashboardFragment.java b/src/com/android/settings/development/DevelopmentSettingsDashboardFragment.java
index 20a80d164e..dac0749e8d 100644
--- a/src/com/android/settings/development/DevelopmentSettingsDashboardFragment.java
+++ b/src/com/android/settings/development/DevelopmentSettingsDashboardFragment.java
@@ -57,7 +57,8 @@ import java.util.List;
 public class DevelopmentSettingsDashboardFragment extends RestrictedDashboardFragment
         implements SwitchBar.OnSwitchChangeListener, OemUnlockDialogHost, AdbDialogHost,
         AdbClearKeysDialogHost, LogPersistDialogHost,
-        BluetoothA2dpHwOffloadRebootDialog.OnA2dpHwDialogConfirmedListener {
+        BluetoothA2dpHwOffloadRebootDialog.OnA2dpHwDialogConfirmedListener,
+        RootAccessDialogHost {
 
     private static final String TAG = "DevSettingsDashboard";
 
@@ -277,31 +278,45 @@ public class DevelopmentSettingsDashboardFragment extends RestrictedDashboardFra
         controller.onA2dpHwDialogConfirmed();
     }
 
-    @Override
-    public void onActivityResult(int requestCode, int resultCode, Intent data) {
-        boolean handledResult = false;
-        for (AbstractPreferenceController controller : mPreferenceControllers) {
-            if (controller instanceof OnActivityResultListener) {
-                // We do not break early because it is possible for multiple controllers to
-                // handle the same result code.
-                handledResult |=
-                        ((OnActivityResultListener) controller).onActivityResult(
-                                requestCode, resultCode, data);
-            }
-        }
-        if (!handledResult) {
-            super.onActivityResult(requestCode, resultCode, data);
-        }
-    }
-
-    @Override
-    protected String getLogTag() {
-        return TAG;
-    }
-
-    @Override
-    public int getHelpResource() {
-        return 0;
+	@Override
+		public void onRootAccessDialogConfirmed() {
+			final RootAccessPreferenceController controller =
+				getDevelopmentOptionsController(RootAccessPreferenceController.class);
+			controller.onRootAccessDialogConfirmed();
+		}
+
+	@Override
+		public void onRootAccessDialogDismissed() {
+			final RootAccessPreferenceController controller =
+				getDevelopmentOptionsController(RootAccessPreferenceController.class);
+			controller.onRootAccessDialogDismissed();
+		}
+
+	@Override
+		public void onActivityResult(int requestCode, int resultCode, Intent data) {
+			boolean handledResult = false;
+			for (AbstractPreferenceController controller : mPreferenceControllers) {
+				if (controller instanceof OnActivityResultListener) {
+					// We do not break early because it is possible for multiple controllers to
+					// handle the same result code.
+					handledResult |=
+						((OnActivityResultListener) controller).onActivityResult(
+							requestCode, resultCode, data);
+				}
+			}
+			if (!handledResult) {
+				super.onActivityResult(requestCode, resultCode, data);
+			}
+		}
+
+	@Override
+		protected String getLogTag() {
+			return TAG;
+		}
+
+	@Override
+		public int getHelpResource() {
+			return 0;
     }
 
     @Override
@@ -455,6 +470,7 @@ public class DevelopmentSettingsDashboardFragment extends RestrictedDashboardFra
         controllers.add(new FreeformWindowsPreferenceController(context));
         controllers.add(new ShortcutManagerThrottlingPreferenceController(context));
         controllers.add(new EnableGnssRawMeasFullTrackingPreferenceController(context));
+		controllers.add(new RootAccessPreferenceController(context, fragment));
         controllers.add(new DefaultLaunchPreferenceController(context, "running_apps"));
         controllers.add(new DefaultLaunchPreferenceController(context, "demo_mode"));
         controllers.add(new DefaultLaunchPreferenceController(context, "quick_settings_tiles"));
diff --git a/src/com/android/settings/development/RootAccessDialogHost.java b/src/com/android/settings/development/RootAccessDialogHost.java
new file mode 100644
index 0000000000..4a31ae8650
--- /dev/null
+++ b/src/com/android/settings/development/RootAccessDialogHost.java
@@ -0,0 +1,33 @@
+/*
+ * Copyright (C) 2018 The LineageOS Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.settings.development;
+
+/**
+ * Interface for RootAccessWarningDialogFragment callbacks.
+ */
+public interface RootAccessDialogHost {
+
+    /**
+     * Called when the user presses ok on the warning dialog.
+     */
+    void onRootAccessDialogConfirmed();
+
+    /**
+     * Called when the user dismisses or cancels the warning dialog.
+     */
+    void onRootAccessDialogDismissed();
+}
diff --git a/src/com/android/settings/development/RootAccessPreferenceController.java b/src/com/android/settings/development/RootAccessPreferenceController.java
new file mode 100644
index 0000000000..cb832d7e00
--- /dev/null
+++ b/src/com/android/settings/development/RootAccessPreferenceController.java
@@ -0,0 +1,137 @@
+/*
+ * Copyright (C) 2018 The LineageOS Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.settings.development;
+
+import android.content.Context;
+import android.os.Build;
+import android.os.ServiceManager;
+import android.os.SystemProperties;
+import android.os.UserManager;
+import android.provider.Settings;
+import android.support.annotation.VisibleForTesting;
+import android.support.v7.preference.ListPreference;
+import android.support.v7.preference.Preference;
+import android.support.v7.preference.PreferenceScreen;
+
+import com.android.settings.R;
+import com.android.settings.Utils;
+import com.android.settings.core.PreferenceControllerMixin;
+import com.android.settingslib.development.DeveloperOptionsPreferenceController;
+
+import java.io.File;
+
+public class RootAccessPreferenceController extends DeveloperOptionsPreferenceController
+        implements Preference.OnPreferenceChangeListener, PreferenceControllerMixin {
+
+    private static final String TAG = "RootAccessPreferenceController";
+    private static final String PREF_KEY = "root_access";
+
+    private static final String ROOT_ACCESS_PROPERTY = "persist.sys.root_access";
+
+    private final DevelopmentSettingsDashboardFragment mFragment;
+    private Object mPendingRootAccessValue;
+
+    public RootAccessPreferenceController(Context context,
+            DevelopmentSettingsDashboardFragment fragment) {
+        super(context);
+
+        mFragment = fragment;
+    }
+
+    @Override
+    public boolean isAvailable() {
+        // User builds don't get root, and eng always gets root
+        return Build.IS_DEBUGGABLE || "eng".equals(Build.TYPE);
+    }
+
+    @Override
+    public String getPreferenceKey() {
+        return PREF_KEY;
+    }
+
+    @Override
+    public void displayPreference(PreferenceScreen screen) {
+        super.displayPreference(screen);
+
+        File file = new File("/system/xbin/su");
+        if (file.exists()) {
+            ((ListPreference) mPreference).setEntries(R.array.root_access_entries);
+            ((ListPreference) mPreference).setEntryValues(R.array.root_access_values);
+        } else {
+            ((ListPreference) mPreference).setEntries(R.array.root_access_entries_adb);
+            ((ListPreference) mPreference).setEntryValues(R.array.root_access_values_adb);
+        }
+
+        updatePreference();
+
+        if (!isAdminUser()) {
+            mPreference.setEnabled(false);
+        }
+    }
+
+    @Override
+    public boolean onPreferenceChange(Preference preference, Object newValue) {
+        if ("0".equals(newValue.toString())) {
+            writeRootAccessOptions(newValue);
+        } else {
+            mPendingRootAccessValue = newValue;
+            RootAccessWarningDialog.show(mFragment);
+        }
+        return true;
+    }
+
+    @Override
+    protected void onDeveloperOptionsSwitchEnabled() {
+        if (isAdminUser()) {
+            mPreference.setEnabled(true);
+        }
+    }
+
+    public void onRootAccessDialogConfirmed() {
+        writeRootAccessOptions(mPendingRootAccessValue);
+    }
+
+    public void onRootAccessDialogDismissed() {
+        updatePreference();
+    }
+
+    private void writeRootAccessOptions(Object newValue) {
+        String oldValue = SystemProperties.get(ROOT_ACCESS_PROPERTY, "0");
+        SystemProperties.set(ROOT_ACCESS_PROPERTY, newValue.toString());
+        if (Integer.valueOf(newValue.toString()) < 2 && !oldValue.equals(newValue)
+                && SystemProperties.getInt("service.adb.root", 0) == 1) {
+            SystemProperties.set("service.adb.root", "0");
+            Settings.Secure.putInt(mContext.getContentResolver(),
+                    Settings.Secure.ADB_ENABLED, 0);
+            Settings.Secure.putInt(mContext.getContentResolver(),
+                    Settings.Secure.ADB_ENABLED, 1);
+        }
+        updatePreference();
+    }
+
+    private void updatePreference() {
+        String value = SystemProperties.get(ROOT_ACCESS_PROPERTY, "0");
+        ((ListPreference) mPreference).setValue(value);
+        ((ListPreference) mPreference).setSummary(mContext.getResources()
+                .getStringArray(R.array.root_access_entries)[Integer.valueOf(value)]);
+    }
+
+    @VisibleForTesting
+    boolean isAdminUser() {
+        return ((UserManager) mContext.getSystemService(Context.USER_SERVICE)).isAdminUser();
+    }
+}
diff --git a/src/com/android/settings/development/RootAccessWarningDialog.java b/src/com/android/settings/development/RootAccessWarningDialog.java
new file mode 100644
index 0000000000..a102c98b0c
--- /dev/null
+++ b/src/com/android/settings/development/RootAccessWarningDialog.java
@@ -0,0 +1,82 @@
+/*
+ * Copyright (C) 2018 The LineageOS Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.settings.development;
+
+import android.app.AlertDialog;
+import android.app.Dialog;
+import android.app.Fragment;
+import android.app.FragmentManager;
+import android.content.DialogInterface;
+import android.os.Bundle;
+
+import com.android.internal.logging.nano.MetricsProto;
+import com.android.settings.R;
+import com.android.settings.core.instrumentation.InstrumentedDialogFragment;
+
+public class RootAccessWarningDialog extends InstrumentedDialogFragment implements
+        DialogInterface.OnClickListener, DialogInterface.OnDismissListener {
+
+    public static final String TAG = "RootAccessWarningDialog";
+
+    public static void show(Fragment host) {
+        final FragmentManager manager = host.getActivity().getFragmentManager();
+        if (manager.findFragmentByTag(TAG) == null) {
+            final RootAccessWarningDialog dialog =
+                    new RootAccessWarningDialog();
+            dialog.setTargetFragment(host, 0 /* requestCode */);
+            dialog.show(manager, TAG);
+        }
+    }
+
+    @Override
+    public int getMetricsCategory() {
+        return MetricsProto.MetricsEvent.TYPE_UNKNOWN;
+    }
+
+    @Override
+    public Dialog onCreateDialog(Bundle savedInstanceState) {
+        return new AlertDialog.Builder(getActivity())
+                .setTitle(R.string.root_access_warning_title)
+                .setMessage(R.string.root_access_warning_message)
+                .setPositiveButton(android.R.string.ok, this /* onClickListener */)
+                .setNegativeButton(android.R.string.cancel, this /* onClickListener */)
+                .create();
+    }
+
+    @Override
+    public void onClick(DialogInterface dialog, int which) {
+        final RootAccessDialogHost host = (RootAccessDialogHost) getTargetFragment();
+        if (host == null) {
+            return;
+        }
+        if (which == DialogInterface.BUTTON_POSITIVE) {
+            host.onRootAccessDialogConfirmed();
+        } else {
+            host.onRootAccessDialogDismissed();
+        }
+    }
+
+    @Override
+    public void onDismiss(DialogInterface dialog) {
+        super.onDismiss(dialog);
+        final RootAccessDialogHost host = (RootAccessDialogHost) getTargetFragment();
+        if (host == null) {
+            return;
+        }
+        host.onRootAccessDialogDismissed();
+    }
+}
-- 
2.17.1

